Polish datetime popover.

jasmussen opened the pull request:
The popover inherited a min-width of 260, which is insufficient, then overrode it just for the edit-post-schedule flow. This means the datetime component looks fine in the sidebar, but no-where else.

This PR changes to use a padding in the component itself, making it work everywhere.

Example of component used outside of sidebar, before:

![Screenshot 2019-11-01 at 12 23 11](https://user-images.githubusercontent.com/1204802/68021673-67ed0980-fca2-11e9-98ed-e219dedf7821.png)


After:

![Screenshot 2019-11-01 at 12 28 45](https://user-images.githubusercontent.com/1204802/68021955-2872ed00-fca3-11e9-98ea-9887c32e0d87.png)

jasmussen committed:
Polish datetime popover.

The popover inherited a min-width of 260, which is insufficient, then overrode it just for the edit-post-schedule flow. This means the datetime component looks fine in the sidebar, but no-where else.

This PR changes the base inherited minwidth to 270, so the component can be used outside of the sidebar.

jorgefilipecosta reviewed (commented):
Hi @jasmussen, on the publish sidebar where the component is not used in a popover, there is a small visual change.
Before:
<img width="281" alt="Screenshot 2019-11-01 at 13 07 55" src="https://user-images.githubusercontent.com/11271197/68026930-f795b680-fca8-11e9-9108-c878f3a8f9b3.png">
After:
<img width="269" alt="Screenshot 2019-11-01 at 13 06 42" src="https://user-images.githubusercontent.com/11271197/68026942-fe242e00-fca8-11e9-8a9a-e9a1c0aedf9d.png">

I just wanted to confirm if that change is ok, or if we should add a rule to avoid that change.

jasmussen commented:
No, good catch, let me see if I can fix that one.

jasmussen commented:
Fixed, thanks.

youknowriad commented on packages/components/src/date-time/style.scss:
```diff
@@ -5,6 +5,8 @@
 /*rtl:end:ignore*/
 
 .components-datetime {
+	padding: $grid-size;
```
Is this needed for all datetime components even outside the modal context? I feel this is specific to the editor usage.

jasmussen replied:
You are right. This component that is included is horrendously messy with arbitrary inline styles. It needs a fresh approach, so I've committed some work in progress and will try a separate PR that goes at it differently.

![Screenshot 2019-11-15 at 09 37 41](https://user-images.githubusercontent.com/1204802/68929228-3ab55680-078c-11ea-9d1f-89c75e0378ce.png)

![Screenshot 2019-11-15 at 09 37 35](https://user-images.githubusercontent.com/1204802/68929231-3d17b080-078c-11ea-8a02-816dfcc69390.png)

┻━┻ ︵ヽ(`Д´)ﾉ︵﻿ ┻━┻o

jasmussen replied:
But it is a good question, though, should it come with a padding or not?

The original version of this PR suggested that _yes_, this component should be born with padding, and if you need the component inline, as in the post-publish, you should override it there. Because otherwise you use the vanilla Popover component with the vanilla DateTime component, and you get this:

![Screenshot 2019-11-15 at 09 46 50](https://user-images.githubusercontent.com/1204802/68929633-05f5cf00-078d-11ea-876e-7db018b58dc5.png)

I can remove some negative margins to fix the horizontal scrollbar:

![Screenshot 2019-11-15 at 09 49 48](https://user-images.githubusercontent.com/1204802/68929791-4e14f180-078d-11ea-9341-10e421481d3d.png)

But the question becomes, if the datetime component should not have padding, should the popover?

I'm going to try something different, but this is context for why I added the padding.

jasmussen replied:
What's causing the horizontal scrollbar in the first place is this:

![Screenshot 2019-11-15 at 09 58 34](https://user-images.githubusercontent.com/1204802/68930389-8d900d80-078e-11ea-9159-bba5c1f84b2d.png)

And it is virtually impossible to refactor that width away because the entire datetime component referenced is reliant on this.

jasmussen committed:
Better code, embrace padding.

The DateTime component as it exists, references a 3rd party library that uses lots of inline styles and hard-coded widths and dimensions.

Having explored various other approaches now, there are only two ways to fix this.

1. Actually add the padding to the component itself. This is the only way to make it work inside popovers when used outside of WordPress.
2. Do a major refactor to include a different datetime library.

In order to move forward, I'm therefore embracing the padding.

jasmussen commented:
Thank you, Riad, for the review in https://github.com/WordPress/gutenberg/pull/18235#pullrequestreview-316875457. 

As you can see from the ensuing discussion I tried my best to work outside the baked in padding. However I failed to do so, and I see no way of fixing this issue without throwing out the calendar 3rd party library entirely and picking a different one. It's just not that good and has so many hardcoded inline styles that are necessary for it to function. 

The good news is, I've applied a bunch of polish to this PR:

![datetime](https://user-images.githubusercontent.com/1204802/68931203-5cb0d800-0790-11ea-99ca-d499b232185e.gif)

Same-height inputs, more cohesive padding and spacing. It looks way better, and it fixes the issue.

The only price is, when you mean to use the datetime component inline, as in the postpublish dialog, you have to unset the left and right paddings. 

I think that's a small price to pay to move forward, without a total refactor. Let me know what you think. I've rebased and squashed this nicely so it should be easy to re-review.

youknowriad commented:
To be fair, I still don't get it. Why instead of overriding the padding for non-popover usage, do the opposite. No padding built-in and override the styles for the popover case only (which is the special case for me)

jasmussen commented:
> To be fair, I still don't get it. Why instead of overriding the padding for non-popover usage, do the opposite. No padding built-in and override the styles for the popover case only (which is the special case for me)

Fair pushback, I suppose I didn't fully explain why I _prefer_ this. It has to do with the following:

<img src="https://user-images.githubusercontent.com/1204802/68930389-8d900d80-078e-11ea-9159-bba5c1f84b2d.png" />

The date component adds a horizontal scrollbar if the popover is smaller than 255px + 2x13px added further up that hierarchy. The min-width for popovers is 260px, defined elsewhere. Which leaves us two ways to avoid a horizontal scrollbar when used outside of the sidebar:

1. Add padding left and right. This increases the effective width of the popover to make room for the 255px + 2x13px, without having to chagne the minimum popover width.
2. Change the minimum popover width to 270 or 280px. 

I can't stress how ugly the code for the date time library is — you kind of have to inspect it to understand why. I mean a third option is to add additional CSS classes or wrappers for any third party that means to use the component inside the popover, and set that class to 280px width. But it's hack upon hack.

youknowriad commented:
I've used this branch and applied this commit 53d2a09c5151dcd70ea655febebc57d208e29c84 to invert the application of the padding. I'm trying to understand what breaks?

jasmussen commented:
> I've used this branch and applied this commit 53d2a09 to invert the application of the padding. I'm trying to understand what breaks?

This happens:

![Screenshot 2019-11-15 at 14 27 15](https://user-images.githubusercontent.com/1204802/68946764-0e153500-07b4-11ea-99bb-23083e2bd9ad.png)

That is how the date-time component looks if you don't apply the following CSS:

```
.edit-post-post-schedule__dialog .components-datetime {
    padding: 8px;
}
```

In other words, if the date time component is used in a popover outside of `edit-post-post-schedule__dialog`, it will have a horizontal scrollbar. Unless every time it's used in that configuration, it has an additional wrapping div with 8px padding applied.

Sure, that's an option — but it feels like it optimizes the date time component to be used inline, vs. used in a popover, and I would suggest the latter is far more likely.

jasmussen commented:
I may be missing something obvious, if that's the case I do apologize. I was up at 5am this morning.

youknowriad commented:
Ok, I understand better, sorry for the back and forth.

I think all components in `@wordpress/components` should be optimized to be used inline. I see the popover as the exception.

That said, I checked deeper and yeah, the issue is the negative margins. I think we can probably remove them and find a way to remove the big padding that calendar has but I'm fine with this PR making an exception to the rule above.

jasmussen commented:
Rebased to see if that makes the checks pass. 

Thanks so much for the reviews, Riad. Every question you've asked is right, and honestly I wanted to address them. But the way the datetime component is built, it really deserves a bigger refactor to make this all possible. The problem with those negative margins is that it's a _horizontally sliding calendar sheet_ that's built in a roundabout way. If we tinker with the margin, this sliding effect breaks when you press "next", which means the styles that are set inline need to be overridden also, and at some point we've reimplemented the whole thing in overriding CSS 😭 — it's a castle made of sand.

But I do believe there is a ticket to revisit the datetime component entirely, that will be a good opportunity to make this better.
